specs/AppNotifications/AppNotificationContentSpec/AppNotificationBuilder-spec.md

```diff
@@ -0,0 +1,1178 @@
+```
+
+XML output:
+
+```xml
+<toast>
+  <visual>
+    <binding template="ToastGeneric">
+      <text>To the left is an AppLogoOverride</text>
+      <image src="https://unsplash.it/64?image=669"
+                placement="appLogoOverride" />
+    </binding>
+  </visual>
+</toast>
+```
+
+## Hero
+
+Prominently displays image within the AppNotification banner and while inside Notification Center.
+
+![Hero Image Example](HeroImageExample.png)
+
+AppNotificationContent code:
+
+```cpp
+AppNotificationContent()
+    .AddText(Text(L"Above is a Hero image"))
+    .SetHeroImage(winrt::Windows::Foundation::Uri(LR"(https://unsplash.it/360/180?image=1043)"))
+    .GetXml();
+```
+
+XML output:
+
+```xml
+<toast>
+  <visual>
+    <binding template="ToastGeneric">
+      <text>Above is a Hero image</text>
+      <image src="https://unsplash.it/360/180?image=1043"
+                placement="hero" />
+    </binding>
+  </visual>
+</toast>
+```
+
+## Cropped AppLogoOverride
+
+For both Inline and AppLogoOverride, developers can set how the image will be cropped.
+
+![Cropped AppLogoOverride Image Example](CroppedAppLogoOverrideExample.png)
+
+AppNotificationContent code:
+
+```cpp
+AppNotificationContent()
+    .AddText(L"To the left is a cropped AppLogoOverride")
+    .SetAppLogoOverride(winrt::Windows::Foundation::Uri(LR"(https://unsplash.it/64?image=669)"), ImageCrop::Circle);
+```
+
+XML output:
+
+```xml
+<toast>
+  <visual>
+    <binding template="ToastGeneric">
+      <text>To the left is a cropped AppLogoOverride</text>
+      <image src="https://unsplash.it/64?image=669"
+                hint-crop="circle"
+                placement="appLogoOverride" />
+    </binding>
+  </visual>
+</toast>
+```
+
+# Audio
+
+The Audio component allows the developer to define a custom audio to play when an AppNotification is
+displayed. The audio file can be defined by string value pointing to an app asset or one of the
+[ms-winsoundevents](https://docs.microsoft.com/uwp/schemas/tiles/toastschema/element-audio).
+
+**WinAppSDK 1.2 `<audio>` Schema**:
+
+```xml
+<audio src? = string
+       loop? = boolean
+       silent? = boolean />
+```
+
+These attributes are abstracted with the Audio component APIs:
+
+-   SetAudio(Windows.Foundation.Uri audioUri)
+-   SetAudio(AppNotficiationSoundEvent appNotficiationSoundEvent)
+-   SetAudio(Windows.Foundation.Uri audioUri, AppNotificationDuration duration)
+-   SetAudio(AppNotficiationSoundEvent appNotficiationSoundEvent, AppNotificationDuration duration)
+-   MuteAudio()
+
+Below are some example usages:
+
+```cpp
+AppNotificationContent()
+    .SetAudio(AppNotficiationSoundEvent::Reminder)
+    GetXml();
+```
+
+XML output:
+
+```xml
+<toast>
+  <visual>
+    <binding template="ToastGeneric">
+    </binding>
+  </visual>
+  <audio src="ms-winsoundevent:Notification.Reminder" />
+</toast>
+```
+
+Developers can define if the audio should be looped using SetAudio(Uri, AppNotificationDuration),
+which can be short or long.
```

Comment:
Sentence doesn't make logical sense. Looping and duration are unrelated concepts in principle. You can have a short looping sound, or a long non-looping sound.

If there is some sort of connection between them, we should say so. "Looped audio must have long duration."

But overloading "Long" to mean "Looped" seems confusing (and possibly even wrong).

The Microsoft.Toolkit.Uwp.Notifications.ToastAudio just calls it Loop and doesn't tie it to duration.

Reply:
The idea was to make it easier for the developers to set a loop value. The developer would have to know to set the duration in the AppNotificationBuilder and we would like to do that for them. What do you think we should do here?

Reply:
Yeah, it is a bit confusing that "Long" == "looping", maybe we have the following...

```csharp
// Non-looping
SetAudio(uri);

// Looping => Setting Loop will automatically ensure at least "short" or "long" duration is specified (defaults to "short" which is the default unless app developer set "long" explicitly)
SetAudio(uri, AppNotificationAudioLooping::Loop);
```

Setting the audio to loop would also ensure the duration property is written in the XML, but the developer doesn't have to know that. They just set looping audio, and if they also want a long notification, they also set the notification to duration long.

Reply:
I wonder also consider if looping audio is something this API should support, without a corresponding _Don't let apps be annoying_ toggle in the UI somewhere.

Reply:
I'm now thinking that since Loop/LoopDuration are not connected, to revert LoopDuration to a Boolean and clarify in the docs to set the duration in the AppNotificationBuilder::SetDuration. What do you think?

Reply:
@pmpurifoy why not just auto-set the duration (to default value)? Since there's no harm in it being set to the default value? Then it "just works" as a developer would expect. Are there any drawbacks to doing that?

* Developer sets loop without anything else -> You create an XML payload and add `duration="short"` (which is the default duration), audio correctly loops as expected, developer is happy
* Developer sets loop and sets duration=long -> You create an XML payload and set `duration="long"`, developer is happy

Reply:
I tested loop="true" without setting the duration and it loops.
I tested loop="true" and duration="short" and it loops.

Both have the same behavior and duration. Since empty duration and duration="short" is the same scenario, we can use Duration::Default without the developer specifying the length. I'm not sure about how we should expose the loop.

Instead of an enum defining loop, I was thinking something like
- SetAudioUriWithLoop
- SetAudioEventWithLoop

AppNotificationDuration::Default is the default and the developer can set the AppNotification to long if they want with AppNotificationDuration::Long. What do you think about this?

Reply:
Awesome to hear that audio looping isn't at all related to the toast duration property!

I'm not a fan of adding `WithLoop` to the methods like `SetAudioUriWithLoop`... I believe we discussed a similar pattern for another one of these builder APIs and agreed that these "properties" should be part of the parameters.

Why not just an additional parameter? What's the drawback of the additional enum aside from there being an additional enum? Btw, I'd also be okay with a boolean, but seemed like API review preferred enums.

```csharp
SetAudioUri(uri, AppNotificationAudioLooping::Loop);
```

Reply:
There is no drawback of adding the enum and I'll add this to the spec, but I was concerned that since the property is loop=bool, we wouldn't be expecting behavior changes like LoopTwice/LoopThreeTimes and the enum would just be an empty enum with one style.

Reply:
What's the use case for playing looping audio with a toast that is no longer visible? How does a user identify what's making noise? What's the case for audio at all? Is this documented somewhere?

Reply:
Audio goes away when the user swipes away or interacts with the toast. Looping just means the notification sound is played repeatedly until the toast goes away. Case for audio is apps can have a unique sound so users will be easily able to identify what app the notification is coming from immediately just based on the sound.

Reply:
Not sure if we explicitly say that the lifetime of the audio is tied to the toast itself anywhere.

Reply:
Implemented AppNotificationAudioLooping

Reply:
> Audio goes away when the user swipes away or interacts with the toast. Looping just means the notification sound is played repeatedly until the toast goes away.

So it is tied to the duration of the toast... Looks like this is happening regardless of what I ask/say so I'm going to bow out and just hope for the best.

Reply:
Audio is tied to the duration of the toast when it first appears, but the audio goes away once the toasts moves to the action center or is activated by the user.